File: app/src/session/tokenStore.js
// KALICI OTURUM DEPOSU — platforma göre EN GÜVENLİ yer seçilir:
//
//   • Android / iOS : expo-secure-store → Android Keystore / iOS Keychain.
//                     Belirteçler cihazın donanım destekli güvenli alanında durur.
//   • Web           : üretimde HttpOnly çerez (sunucu yönetir; JS okuyamaz),
//                     geliştirmede localStorage (yalnız kendi makinen).
//
// ŞİFRE HİÇBİR ZAMAN CİHAZDA SAKLANMAZ — yalnız belirteçler saklanır.
//
// ANAHTAR ADLARI: 'sportoto.token' adı KORUNUR (mevcut kurulumlar bu adla
// kayıtlı). Mobilde tek seferlik göç yapılır: AsyncStorage'daki eski belirteç
// SecureStore'a taşınır ve düz depodan SİLİNİR.
import { Platform } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import * as SecureStore from 'expo-secure-store';
import { API_BASE } from '../config';

const KEY_TOKEN = 'sportoto.token';        // AD DEĞİŞMEZ (mevcut veriyle uyum)
const KEY_REFRESH = 'sportoto.refresh';    // yeni: yenileme anahtarı
const KEY_SESSION = 'sportoto.session';    // yeni: oturum kimliği
const KEY_DEVICE = 'sportoto.device';      // yeni: kalıcı cihaz kimliği (rastgele)

const isWeb = Platform.OS === 'web';

// SecureStore anahtarları yalnız [A-Za-z0-9._-] içerebilir — mevcut adlar uygun.
async function secureGet(k) { try { return await SecureStore.getItemAsync(k); } catch { return null; } }
async function secureSet(k, v) { try { if (v) await SecureStore.setItemAsync(k, v); else await SecureStore.deleteItemAsync(k); } catch { /* okuma tarafı null görür */ } }

function webGet(k) { try { return typeof localStorage !== 'undefined' ? localStorage.getItem(k) : null; } catch { return null; } }
function webSet(k, v) { try { if (typeof localStorage === 'undefined') return; if (v) localStorage.setItem(k, v); else localStorage.removeItem(k); } catch {} }

async function get(k) { return isWeb ? webGet(k) : secureGet(k); }
async function set(k, v) { return isWeb ? webSet(k, v) : secureSet(k, v); }

/**
 * Web'de ÇEREZ MODU: sayfa ile API aynı origin'deyse (üretimde backend web
 * derlemesini kendisi sunar) belirteçler HttpOnly çerezle taşınır — localStorage'a
 * hiç yazılmaz (XSS bir şey çalamaz). Geliştirmede (farklı port) Bearer kullanılır.
 */
export function cookieModeAvailable() {
  if (!isWeb || typeof location === 'undefined') return false;
  if (!API_BASE) return true; // boş taban = aynı origin (yayın web derlemesi)
  try { return new URL(API_BASE).origin === location.origin; } catch { return false; }
}

/** Mobil: eski AsyncStorage belirtecini SecureStore'a tek seferlik taşır. */
async function migrateLegacyToken() {
  if (isWeb) return;
  try {
    const inSecure = await secureGet(KEY_TOKEN);
    if (inSecure) return;
    const legacy = await AsyncStorage.getItem(KEY_TOKEN);
    if (legacy) {
      await secureSet(KEY_TOKEN, legacy);
      await AsyncStorage.removeItem(KEY_TOKEN); // düz metin depoda İZ KALMAZ
    }
  } catch { /* göç başarısızsa kullanıcı yalnızca yeniden giriş yapar */ }
}

/** Uygulama açılışında kalıcı oturumu yükler. */
export async function loadPersisted() {
  await migrateLegacyToken();
  const [token, refreshToken, sessionId] = await Promise.all([
    get(KEY_TOKEN), get(KEY_REFRESH), get(KEY_SESSION),
  ]);
  return { token, refreshToken, sessionId };
}

/** Oturumu kalıcılaştırır. Çerez modunda belirteçler DEPOLANMAZ (çerezdedir). */
export async function persistSession({ token, refreshToken, sessionId, cookieMode }) {
  if (cookieMode) {
    await Promise.all([set(KEY_TOKEN, null), set(KEY_REFRESH, null), set(KEY_SESSION, sessionId || null)]);
    return;
  }
  await Promise.all([
    set(KEY_TOKEN, token || null),
    set(KEY_REFRESH, refreshToken || null),
    set(KEY_SESSION, sessionId || null),
  ]);
}

export async function clearPersisted() {
  await Promise.all([set(KEY_TOKEN, null), set(KEY_REFRESH, null), set(KEY_SESSION, null)]);
}

/** Kalıcı, rastgele cihaz kimliği (kişisel veri DEĞİLDİR; cihaz listesi için). */
export async function getOrCreateDeviceId() {
  let id = await get(KEY_DEVICE);
  if (!id) {
    id = (typeof crypto !== 'undefined' && crypto.randomUUID)
      ? crypto.randomUUID()
      : `d-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 10)}`;
    await set(KEY_DEVICE, id);
  }
  return id;
}

/** Cihaz listesinde görünecek ad — kişisel veri içermez. */
export function deviceLabel() {
  if (isWeb) {
    const ua = typeof navigator !== 'undefined' ? navigator.userAgent : '';
    const browser = /Edg\//.test(ua) ? 'Edge' : /Chrome\//.test(ua) ? 'Chrome' : /Firefox\//.test(ua) ? 'Firefox' : /Safari\//.test(ua) ? 'Safari' : 'Tarayıcı';
    return `Web · ${browser}`;
  }
  return Platform.OS === 'android' ? 'Android' : Platform.OS === 'ios' ? 'iPhone/iPad' : Platform.OS;
}

export function platformName() { return Platform.OS; }

File: lib/core/session/token_store.dart
// KAYNAK: app/src/session/tokenStore.js — çeviri.
//
// KALICI OTURUM DEPOSU — platforma göre EN GÜVENLİ yer seçilir:
//
//   • Android / iOS : flutter_secure_storage → Android Keystore (EncryptedShared
//                     Preferences) / iOS Keychain. Kaynakta expo-secure-store
//                     aynı iki mekanizmayı kullanıyordu.
//   • Web           : üretimde HttpOnly çerez (sunucu yönetir; JS okuyamaz),
//                     geliştirmede tarayıcı deposu (yalnız kendi makinen).
//
// ŞİFRE HİÇBİR ZAMAN CİHAZDA SAKLANMAZ — yalnız belirteçler saklanır.
//
// ANAHTAR ADLARI KAYNAKTAN AYNEN KORUNDU ('sportoto.token' vb.).
//
// ────────────────────────────────────────────────────────────────────────────
// BİLEREK ÇEVRİLMEYEN TEK PARÇA: `migrateLegacyToken()`
//
// Kaynakta AsyncStorage'daki eski düz-metin belirteci SecureStore'a taşıyan bir
// tek-seferlik göç vardı. Buraya taşınmadı, çünkü taşınması İMKÂNSIZ: Expo'nun
// SecureStore'u ile flutter_secure_storage aynı anahtar takma adlarını ve aynı
// depoyu kullanmaz. Flutter derlemesi cihaza kurulduğunda RN sürümünün
// belirtecini OKUYAMAZ.
//
// KULLANICIYA GÖRÜNEN SONUÇ: mevcut kullanıcılar Flutter sürümüne geçtiğinde
// BİR KEZ yeniden giriş yapar. Sessizce olmaması için burada yazılıdır.

import 'dart:convert';
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import '../network/api_config.dart';

const String _keyToken = 'sportoto.token'; // AD DEĞİŞMEZ
const String _keyRefresh = 'sportoto.refresh';
const String _keySession = 'sportoto.session';
const String _keyDevice = 'sportoto.device';

// Varsayılan AndroidOptions BİLEREK kullanılıyor: flutter_secure_storage 11'de
// varsayılan zaten donanım destekli Keystore'dur (AES/GCM veri + RSA-OAEP
// anahtar sarma, API 23+). Eski sürümlerdeki `encryptedSharedPreferences: true`
// bayrağı kaldırıldı çünkü artık ondan daha güçlüsü varsayılan.
// Biyometri ZORUNLU KILINMADI: kaynakta belirteç okumak biyometri istemez;
// biyometrik kilit ayrı ve kullanıcı isteğine bağlı bir katmandır
// (app/src/security/biometricLock.js → Adım 4).
const _storage = FlutterSecureStorage();

Future<String?> _get(String k) async {
  try {
    return await _storage.read(key: k);
  } catch (_) {
    return null;
  }
}

Future<void> _set(String k, String? v) async {
  try {
    if (v != null && v.isNotEmpty) {
      await _storage.write(key: k, value: v);
    } else {
      await _storage.delete(key: k);
    }
  } catch (_) {
    /* okuma tarafı null görür */
  }
}

/// Web'de ÇEREZ MODU: sayfa ile API aynı origin'deyse (üretimde backend web
/// derlemesini kendisi sunar) belirteçler HttpOnly çerezle taşınır — tarayıcı
/// deposuna hiç yazılmaz. Geliştirmede (farklı port) Bearer kullanılır.
bool cookieModeAvailable() {
  if (!kIsWeb) return false;
  // boş taban = aynı origin (yayın web derlemesi)
  if (apiBase.isEmpty) return true;
  try {
    final api = Uri.parse(apiBase);
    final here = Uri.base;
    return api.scheme == here.scheme &&
        api.host == here.host &&
        api.port == here.port;
  } catch (_) {
    return false;
  }
}

class PersistedSession {
  const PersistedSession({this.token, this.refreshToken, this.sessionId});
  final String? token;
  final String? refreshToken;
  final String? sessionId;
}

/// Uygulama açılışında kalıcı oturumu yükler.
Future<PersistedSession> loadPersisted() async {
  final values = await Future.wait([
    _get(_keyToken),
    _get(_keyRefresh),
    _get(_keySession),
  ]);
  return PersistedSession(
    token: values[0],
    refreshToken: values[1],
    sessionId: values[2],
  );
}

/// Oturumu kalıcılaştırır. Çerez modunda belirteçler DEPOLANMAZ (çerezdedir).
Future<void> persistSession({
  String? token,
  String? refreshToken,
  String? sessionId,
  bool cookieMode = false,
}) async {
  if (cookieMode) {
    await Future.wait([
      _set(_keyToken, null),
      _set(_keyRefresh, null),
      _set(_keySession, sessionId),
    ]);
    return;
  }
  await Future.wait([
    _set(_keyToken, token),
    _set(_keyRefresh, refreshToken),
    _set(_keySession, sessionId),
  ]);
}

Future<void> clearPersisted() async {
  await Future.wait([
    _set(_keyToken, null),
    _set(_keyRefresh, null),
    _set(_keySession, null),
  ]);
}

/// Kalıcı, rastgele cihaz kimliği (kişisel veri DEĞİLDİR; cihaz listesi için).
Future<String> getOrCreateDeviceId() async {
  var id = await _get(_keyDevice);
  if (id == null || id.isEmpty) {
    id = _randomUuidV4();
    await _set(_keyDevice, id);
  }
  return id;
}

/// Kaynakta `crypto.randomUUID()` kullanılıyordu. Dart'ta karşılığı yok;
/// kriptografik rastgelelikten RFC 4122 v4 üretilir (zaman damgasına dayanan
/// zayıf yedek KULLANILMAZ — cihaz kimliği tahmin edilebilir olmamalı).
String _randomUuidV4() {
  final rnd = Random.secure();
  final bytes = List<int>.generate(16, (_) => rnd.nextInt(256));
  bytes[6] = (bytes[6] & 0x0f) | 0x40; // sürüm 4
  bytes[8] = (bytes[8] & 0x3f) | 0x80; // varyant 10x
  final hex = bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  return '${hex.substring(0, 8)}-${hex.substring(8, 12)}-'
      '${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20)}';
}

/// Cihaz listesinde görünecek ad — kişisel veri içermez.
String deviceLabel() {
  if (kIsWeb) {
    // Kaynakta userAgent'tan tarayıcı adı çıkarılıyordu. Dart'ta userAgent'a
    // ancak dart:html ile ulaşılır ve o mobil derlemeyi kırar; bu yüzden web
    // tarafı Adım 5'te (web derleme kontrolü) koşullu içe aktarma ile
    // tamamlanacak. Şimdilik ad kaynaktaki genel karşılığa düşer.
    return 'Web · Tarayıcı';
  }
  return switch (defaultTargetPlatform) {
    TargetPlatform.android => 'Android',
    TargetPlatform.iOS => 'iPhone/iPad',
    TargetPlatform.macOS => 'macOS',
    TargetPlatform.windows => 'Windows',
    TargetPlatform.linux => 'Linux',
    TargetPlatform.fuchsia => 'Fuchsia',
  };
}

/// KAYNAK: `Platform.OS` — 'android' | 'ios' | 'web' gibi SABİT değerler.
///
/// Burada enum'un `.name` getter'ı KULLANILMAZ. Yayın derlemesi `--obfuscate`
/// ile karartıldığında tip/alan adlarına dayanan her şey güvenilmez olur;
/// sunucuya giden cihaz adı da öyle. Açık eşleme hem karartmadan etkilenmez
/// hem kaynaktaki değerlerle birebir aynıdır.
String platformName() {
  if (kIsWeb) return 'web';
  return switch (defaultTargetPlatform) {
    TargetPlatform.android => 'android',
    TargetPlatform.iOS => 'ios',
    TargetPlatform.macOS => 'macos',
    TargetPlatform.windows => 'windows',
    TargetPlatform.linux => 'linux',
    TargetPlatform.fuchsia => 'fuchsia',
  };
}

/// Base64 yardımcıları — JWT çözümlemesi gereken yerlerde (oturum süresi
/// göstergesi) kullanılır. Doğrulama YAPMAZ: imza denetimi sunucudadır.
Map<String, dynamic>? decodeJwtPayload(String? jwt) {
  if (jwt == null) return null;
  final parts = jwt.split('.');
  if (parts.length != 3) return null;
  try {
    final normalized = base64Url.normalize(parts[1]);
    return jsonDecode(utf8.decode(base64Url.decode(normalized)))
        as Map<String, dynamic>;
  } catch (_) {
    return null;
  }
}
